assets/src/apps/authoring/store/groups/layouts/deck/actions/validate.ts

```diff
@@ -421,6 +425,51 @@ export const diagnosePage = (page: any, allActivities: any[], sequence: any[]) =
   return errors;
 };
 
+const validateLessonVariables = (page: any) => {
+  const allNames = page.custom.variables.map((v: any) => v.name);
+  // variables can and will ref previous ones
+  // they will reference them "globally" so need to track the above
+  // in order to prepend the "variables" namespace
+  const statements: string[] = page.custom.variables
+    .map((v: any) => {
+      if (!v.name || !v.expression) {
+        return '';
+      }
+      let expr = v.expression;
+      allNames.forEach((name: string) => {
+        const regex = new RegExp(`{${name}}`, 'g');
+        expr = expr.replace(regex, `{variables.${name}}`);
+      });
+
+      const stmt = { expression: `let {variables.${v.name.trim()}} = ${expr};`, name: v.name };
+      return stmt;
+    })
+    .filter((s: any) => s);
+  // execute each sequentially in case there are errors (missing functions)
+  const broken: any[] = [];
+  statements.forEach((statement: any) => {
+    try {
+      const result = evalScript(statement.expression, defaultGlobalEnv);
```

Comment:
for this one you need to do a new environment every execution, otherwise the undefined will become defined on subsequent runs
```
const testEnv = new Environment();
  const broken: any[] = [];
  statements.forEach((statement: any) => {
    try {
      const result = evalScript(statement.expression, testEnv);
```

Reply:
We can talk about it on today's TB call